src/record_ts.cpp

```diff
@@ -0,0 +1,1973 @@
+                    uintptr_t size = 0;
+                    std::memcpy(&size, &launch_size, sizeof(uint32_t));
+                    kernel_params[0] = (void *) size;
+                }
+
+                if (!dry_run) {
+                    jitc_log(LogLevel::Debug,
+                             "replay(): launching kernel %u %016llx [%u]%s",
+                             n_kernels++,
+                             (unsigned long long) op.kernel.hash.high64,
+                             launch_size, op.uses_optix ? " uses optix" : "");
+
+                    std::vector<uint32_t> kernel_param_ids;
+                    std::vector<uint32_t> kernel_calls;
+                    Kernel kernel = op.kernel.kernel;
+                    if (op.uses_optix) {
+                        uses_optix    = true;
+                        ts->optix_sbt = op.sbt;
+                    }
+                    ts->launch(kernel, op.kernel.key, op.kernel.hash,
+                               launch_size, &kernel_params, &kernel_param_ids);
+                    if (op.uses_optix)
+                        uses_optix = false;
+                }
+
+            };
+            break;
+            case OpType::Barrier: {
+                ProfilerPhase profiler(pr_barrier);
+                if (!dry_run)
+                    ts->barrier();
+            };
+            break;
+            case OpType::MemsetAsync: {
+                ProfilerPhase profiler(pr_memset_async);
+
+                uint32_t dependency_index = op.dependency_range.first;
+
+                ParamInfo ptr_info = this->dependencies[dependency_index];
+
+                ReplayVariable &ptr_var = replay_variables[ptr_info.slot];
+                ptr_var.alloc(backend, op.size, op.input_size);
+
+                jitc_log(LogLevel::Debug,
+                         "replay(): MemsetAsync -> s%u [%zu], "
+                         "op.input_size=%zu",
+                         ptr_info.slot, op.size, op.input_size);
+
+                uint32_t size = ptr_var.size(op.input_size);
+
+                if (!dry_run)
+                    ts->memset_async(ptr_var.data, size, op.input_size,
+                                     &op.data);
+            };
+            break;
+            case OpType::ReduceExpanded: {
+                ProfilerPhase profiler(pr_reduce_expanded);
+
+                jitc_log(LogLevel::Debug, "replay(): ReduceExpand");
+
+                uint32_t dependency_index = op.dependency_range.first;
+                ParamInfo data_info = this->dependencies[dependency_index];
+
+                ReplayVariable &data_var = replay_variables[data_info.slot];
+
+                VarType vt       = data_info.vtype;
+                ReduceOp rop     = op.rtype;
+                uint32_t size    = data_var.size(vt);
+                uint32_t tsize   = type_size[(uint32_t) vt];
+                uint32_t workers = pool_size() + 1;
+
+                uint32_t replication_per_worker =
```

Comment:
Good idea. Since it is only used for the LLVM backend, it could be a global function like `jitc_llvm_expand_replication_factor` or something like that.